dyno/microcontroller/Session.ino

```diff
@@ -4,108 +4,156 @@
+{
   m_pot_pin = POT_PIN;
-  for (int i = 0; i < NUM_MAGNETS; i++) {
+  for (int i = 0; i < NUM_MAGNETS; i++)
+  {
     m_drv_pins[i] = DRV_PINS[i];
     m_pwms[i] = 0;
   }
   memset(m_com_buff, 0, sizeof(m_com_buff));
   m_com_buff_count = 0;
   m_drive_w_pot = drive_w_pot;
-  if (paused) m_paused_flag = 1;
-  else m_paused_flag = 0;
+  if (paused)
+    m_paused_flag = 1;
+  else
+    m_paused_flag = 0;
 }
 
-void Session::init_temp_sense(const int *TEMP_PINS) {
-  for (int i = 0; i < NUM_TEMP_SENSE; i++) {
+void Session::init_temp_sense(const int *TEMP_PINS)
+{
+  for (int i = 0; i < NUM_TEMP_SENSE; i++)
+  {
     m_temp_pins[i] = TEMP_PINS[i];
   }
 }
 
-void Session::process_command() {
-  if (m_com_buff_count == 1) {
+bool Session::process_command()
+{
+  auto command_parsed = false;
+  if (m_com_buff_count == 1)
+  {
     //easy parsing for single char commands
-    if (m_com_buff[0] == 'R') {
+    if (m_com_buff[0] == 'R')
+    {
       //resume
       m_paused_flag = false;
+      command_parsed = true;
     }
-    if (m_com_buff[0] == 'P') {
+    if (m_com_buff[0] == 'P')
+    {
       //pause
       m_paused_flag = true;
+      command_parsed = true;
     }
-  } else { //multi-char commands
+  }
+  else
+  { //multi-char commands
     //extract command prefix
     String command = "";
     int i;
-    for (i = 0; i < m_com_buff_count; i++) {
-      if (m_com_buff[i] != '=') command += m_com_buff[i];
-      else {i++; break;}
+    for (i = 0; i < m_com_buff_count; i++)
+    {
+      if (m_com_buff[i] != '=')
+        command += m_com_buff[i];
+      else
+      {
+        i++;
+        break;
+      }
     }
 
     //handle commands
-    if (command == "MAGNET") {
+    if (command == "MAGNET")
+    {
       //assumed format:
-      //MAGNET=1:240,3:000, 
-      while (i < m_com_buff_count) {
+      //MAGNET=1:240,3:000,
+      while (i < m_com_buff_count)
+      {
         //still more to be parsed
-        String pwm_str = String(m_com_buff[i+2]) + String(m_com_buff[i+3]) + String(m_com_buff[i+4]);
+        String pwm_str = String(m_com_buff[i + 2]) + String(m_com_buff[i + 3]) + String(m_com_buff[i + 4]);
         m_pwms[m_com_buff[i] - '0'] = pwm_str.toInt();
         i += 6;
       }
-    } 
+      command_parsed = true;
+    }
   }
   //command stored in 256 byte buffer => can implement multi-char commands too
+
+  return command_parsed;
 }
 
-void Session::check_for_command() {
+void Session::check_for_command()
+{
   //buffer incoming data until ; indicates end of command
   char incomingByte;
-  while (Serial.available() > 0) {
+  while (Serial.available() > 0)
+  {
     incomingByte = Serial.read();
-    if (incomingByte == ';') {
-      process_command();
+    m_serial_stats.m_chars_received += 1;
+    if (incomingByte == ';')
+    {
+      m_serial_stats.m_commands_received += 1;
+      if (process_command())
+      {
+        m_serial_stats.m_commands_parsed += 1;
+        Serial.write("ack;");
```

Comment:
@ashvinnagarajan not sure how you'd prefer to receive the ACK on the panda side, so let me know if you'd rather I drop the ; or use something entirely different here.

Reply:
For the ack, what do you think about using HTTP error codes? something like ack=200; for a success and ack=400; if the request fails?

Reply:
Or even a binary 0 and 1 would work!

Reply:
I think I prefer `ack=200;` so that we can still add other messages going from the microcontroller to the panda. I like the error code idea!